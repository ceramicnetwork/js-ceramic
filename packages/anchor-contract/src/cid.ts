import type { EventFragment } from '@ethersproject/abi'
import type { Log, TransactionResponse } from '@ethersproject/providers'
import { create as createMultihash } from 'multiformats/hashes/digest'
import { CID } from 'multiformats/cid'
import { fromString } from 'uint8arrays'

import { contractInterface } from './interface.js'

const SHA256_CODE = 0x12
const DAG_CBOR_CODE = 0x71
const V1_PROOF_TYPE = 'f(bytes32)'

export function createCidFromHexValue(value: string): CID {
  const multihash = createMultihash(SHA256_CODE, fromString(value.slice(2), 'base16'))
  return CID.create(1, DAG_CBOR_CODE, multihash)
}

export function getCidFromAnchorEventLog(log: Log): CID {
  const decodedArgs = contractInterface.decodeEventLog(
    contractInterface.events['DidAnchor'] as EventFragment,
    log.data
  )
  return createCidFromHexValue(decodedArgs[1])
}

const getCidFromV0Transaction = (txResponse: TransactionResponse): CID => {
  const withoutPrefix = txResponse.data.replace(/^(0x0?)/, '')
  return CID.decode(fromString(withoutPrefix.slice(1), 'base16'))
}

const getCidFromV1Transaction = (txResponse: TransactionResponse): CID => {
  const decodedArgs = contractInterface.decodeFunctionData('anchorDagCbor', txResponse.data)
  return createCidFromHexValue(decodedArgs[0])
}

/**
 * Parses the transaction data to recover the CID.
 * @param txType transaction type of the anchor proof. Currently support `raw` and `f(bytes32)`
 * @param txResponse the retrieved transaction from the ethereum blockchain
 * @returns
 */
export const getCidFromTransaction = (txType: string, txResponse: TransactionResponse): CID => {
  if (txType === V1_PROOF_TYPE) {
    return getCidFromV1Transaction(txResponse)
  } else {
    return getCidFromV0Transaction(txResponse)
  }
}
