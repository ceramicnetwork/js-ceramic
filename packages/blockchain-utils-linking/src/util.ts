export interface LinkProof {
  version: number;
  message: string;
  signature: string;
  account: string;
  did?: string;
  timestamp?: number;
  address?: string;
  type?: string;
  chainId?: number;
}

export interface RpcMessage {
  jsonrpc: string;
  id: number;
  method: string;
  params: any;
}

export interface ConsentMessage {
  message: string;
  timestamp?: number;
}

export function getConsentMessage(
  did: string,
  addTimestamp = true
): ConsentMessage {
  const res: any = {
    message: "Create a new account link to your identity." + "\n\n" + did,
  };
  if (addTimestamp) {
    res.timestamp = Math.floor(Date.now() / 1000);
    res.message += " \n" + "Timestamp: " + res.timestamp;
  }
  return res;
}

export function encodeRpcMessage(method: string, params?: any): any {
  return {
    jsonrpc: "2.0",
    id: 1,
    method,
    params,
  };
}
