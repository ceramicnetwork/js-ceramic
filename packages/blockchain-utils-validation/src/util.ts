import { AccountId } from 'caip'

export function normalizeAccountId(accountId: string | AccountId): AccountId {
  // Handle legacy CAIP-10 representation
  if (typeof accountId === 'string' && accountId.includes('@')) {
    const [address, chainId] = accountId.split('@')
    if (!address || !chainId) {
      throw new Error(`Invalid accountId provided`)
    }

    return new AccountId({ address, chainId })
  }
  return new AccountId(accountId)
}
