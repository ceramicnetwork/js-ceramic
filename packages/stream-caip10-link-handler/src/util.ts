// NOTE: This differs from `normalizeAccountId` in `stream-caip10-link`
// This converts from CAIP-10 repr to the legacy repr
// The former converts from legacy to the current repr
export function normalizeAccountId(accountId: string): string {
  if (!accountId.includes('@')) {
    const _accountSplit = accountId.split(':')
    const address = _accountSplit.pop()
    const chainId = _accountSplit.join(':')
    return `${address}@${chainId}`
  }
  return accountId
}
