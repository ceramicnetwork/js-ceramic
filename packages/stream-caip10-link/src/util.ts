import { AccountId } from 'caip'

export function normalizeAccountId(accountId: string | AccountId): AccountId {
  if (typeof accountId === 'string' && accountId.includes('@')) {
    const [address, chainId] = accountId.split('@')
    if (!address || !chainId) {
      throw new Error(`Invalid accountId provided`)
    }

    return new AccountId({ address, chainId })
  }
  return new AccountId(accountId)
}
