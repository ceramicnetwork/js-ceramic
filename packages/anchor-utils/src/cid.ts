import type { EventFragment } from '@ethersproject/abi'
import type { Log, TransactionResponse } from '@ethersproject/providers'
import { create as createMultihash } from 'multiformats/hashes/digest'
import { CID } from 'multiformats/cid'
import { fromString } from 'uint8arrays'

import { contractInterface } from './interface.js'

export const SHA256_CODE = 0x12
export const KECCAK_256_CODE = 0x1b
export const DAG_CBOR_CODE = 0x71
export const ETH_TX_CODE = 0x93

export function createCidFromHexValue(hexValue: string): CID {
  const multihash = createMultihash(SHA256_CODE, fromString(hexValue.slice(2), 'base16'))
  return CID.create(1, DAG_CBOR_CODE, multihash)
}

export function getCidFromAnchorEventLog(log: Log): CID {
  const decodedArgs = contractInterface.decodeEventLog(
    contractInterface.events['DidAnchor(address,bytes32)'] as EventFragment,
    log.data
  )
  return createCidFromHexValue(decodedArgs[1])
}

const getCidFromV0Transaction = (txResponse: TransactionResponse): CID => {
  const withoutPrefix = txResponse.data.replace(/^(0x0?)/, '')
  return CID.decode(fromString(withoutPrefix.slice(1), 'base16'))
}

const getCidFromV1Transaction = (txResponse: TransactionResponse): CID => {
  const decodedArgs = contractInterface.decodeFunctionData('anchorDagCbor', txResponse.data)
  return createCidFromHexValue(decodedArgs[0])
}

const V1_PROOF_TYPE = 'f(bytes32)'

/**
 * Parses the transaction data to recover the CID.
 * @param txType transaction type of the anchor proof. Currently support `raw` and `f(bytes32)`
 * @param txResponse the retrieved transaction from the ethereum blockchain
 * @returns
 */
export const getCidFromTransaction = (txType: string, txResponse: TransactionResponse): CID => {
  if (txType === V1_PROOF_TYPE) {
    return getCidFromV1Transaction(txResponse)
  } else {
    return getCidFromV0Transaction(txResponse)
  }
}

/**
 * Converts ETH address to CID
 * @param hash - ETH hash
 */
export function convertEthHashToCid(hexValue: string): CID {
  const multihash = createMultihash(KECCAK_256_CODE, fromString(hexValue.slice(2), 'base16'))
  return CID.create(1, ETH_TX_CODE, multihash)
}
