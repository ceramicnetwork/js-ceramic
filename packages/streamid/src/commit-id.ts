import { CID } from 'multiformats/cid'
import { base36 } from 'multiformats/bases/base36'
import { StreamType } from './stream-type.js'
import varint from 'varint'
import { concat as uint8ArrayConcat } from 'uint8arrays'
import { Memoize } from 'typescript-memoize'
import { STREAMID_CODEC } from './constants.js'
import { StreamID } from './stream-id.js'
import type { StreamRef } from './stream-ref.js'
import { tryCatch } from './try-catch.util.js'
import * as parsing from './stream-ref-parsing.js'

export class InvalidCommitIDBytesError extends Error {
  constructor(bytes: Uint8Array) {
    super(
      `Error while parsing CommitID from bytes ${base36.encode(
        bytes
      )}: no commit information provided`
    )
  }
}

export class InvalidCommitIDStringError extends Error {
  constructor(input: string) {
    super(`Error while parsing CommitID from string ${input}: no commit information provided`)
  }
}

/**
 * Parse CommitID from bytes representation.
 *
 * @param bytes - bytes representation of CommitID.
 * @throws error on invalid input
 * @see CommitID#bytes
 */
function fromBytes(bytes: Uint8Array): CommitID {
  const parsed = parsing.fromBytes(bytes, 'CommitID')
  if (parsed.kind === 'commit-id') {
    return new CommitID(parsed.type, parsed.genesis, parsed.commit)
  }
  throw new InvalidCommitIDBytesError(bytes)
}

/**
 * Same as fromBytes, but returns an Error instance rather than throwing if there is a problem
 * with the input.
 * Note that some exceptions can still be thrown in certain cases, if they come from lower-level
 * libraries like multibase, for example.
 * @param bytes
 */
function fromBytesNoThrow(bytes: Uint8Array): CommitID | Error {
  return tryCatch(() => fromBytes(bytes))
}

/**
 * Parse CommitID from string representation.
 *
 * @param input - string representation of CommitID, be it base36-encoded string or URL.
 * @see CommitID#toString
 * @see CommitID#toUrl
 */
function fromString(input: string): CommitID {
  const parsed = parsing.fromString(input, 'CommitID')
  if (parsed.kind === 'commit-id') {
    return new CommitID(parsed.type, parsed.genesis, parsed.commit)
  }
  throw new InvalidCommitIDStringError(input)
}

/**
 * Same as fromString, but returns an Error instance rather than throwing if there is a problem
 * with the input.
 * Note that some exceptions can still be thrown in certain cases, if they come from lower-level
 * libraries like multibase, for example.
 * @param input
 */
function fromStringNoThrow(input: string): CommitID | Error {
  return tryCatch(() => fromString(input))
}

const TAG = Symbol.for('@ceramicnetwork/streamid/CommitID')

/**
 * Construct new CommitID for a given stream and commit
 */
function make(stream: StreamID, commit: CID | string | number): CommitID {
  return new CommitID(stream.type, stream.cid, commit)
}

/**
 * Commit identifier, includes type, genesis CID, commit CID.
 *
 * Encoded as `<multibase-prefix><multicodec-streamid><type><genesis-cid-bytes><commit-cid-bytes>`.
 *
 * String representation is base36-encoding of the bytes above.
 */
export class CommitID implements StreamRef {
  protected readonly _tag = TAG

  readonly #type: number
  readonly #cid: CID
  readonly #commit: CID | null // null ‚âù genesis commit

  static fromBytes = fromBytes
  static fromBytesNoThrow = fromBytesNoThrow
  static fromString = fromString
  static fromStringNoThrow = fromStringNoThrow
  static make = make

  // WORKAROUND. Weird replacement for Symbol.hasInstance due to
  // this old bug in Babel https://github.com/babel/babel/issues/4452
  // which is used by CRA, which is widely popular.
  static isInstance(instance: any): instance is CommitID {
    return typeof instance === 'object' && '_tag' in instance && instance._tag === TAG
  }

  /**
   * Create a new StreamID.
   *
   * @param type
   * @param {CID|string}         cid
   * @param {CID|string}         commit CID. Pass '0', 0, or omit the value as shorthand for the genesis commit.
   *
   * @example
   * new StreamID(<type>, <CID>|<cidStr>)
   * new StreamID(<type>, <CID>|<cidStr>, <CommitCID>|<CommitCidStr>)
   */
  constructor(
    type: string | number,
    cid: CID | string,
    commit: CID | string | number | null = null
  ) {
    if (!type && type !== 0) throw new Error('constructor: type required')
    if (!cid) throw new Error('constructor: cid required')
    this.#type = typeof type === 'string' ? StreamType.codeByName(type) : type
    this.#cid = typeof cid === 'string' ? CID.parse(cid) : cid
    this.#commit = parsing.parseCommit(this.#cid, commit)
  }

  /**
   * Construct StreamID, no commit information included
   */
  @Memoize()
  get baseID(): StreamID {
    return new StreamID(this.#type, this.#cid)
  }

  /**
   * Stream type code
   */
  get type(): number {
    return this.#type
  }

  /**
   * Stream type name
   */
  @Memoize()
  get typeName(): string {
    return StreamType.nameByCode(this.#type)
  }

  /**
   * Genesis CID
   */
  get cid(): CID {
    return this.#cid
  }

  /**
   * Commit CID
   */
  @Memoize()
  get commit(): CID {
    return this.#commit || this.#cid
  }

  /**
   * Bytes representation
   */
  @Memoize()
  get bytes(): Uint8Array {
    const codec = varint.encode(STREAMID_CODEC)
    const type = varint.encode(this.type)

    const commitBytes = this.#commit?.bytes || new Uint8Array([0])
    return uint8ArrayConcat([codec, type, this.cid.bytes, commitBytes])
  }

  /**
   * Compare equality with another CommitID.
   */
  equals(other: CommitID): boolean {
    return (
      this.type === other.type && this.cid.equals(other.cid) && this.commit.equals(other.commit)
    )
  }

  /**
   * Encode the CommitID into a string.
   */
  @Memoize()
  toString(): string {
    return base36.encode(this.bytes)
  }

  /**
   * Encode the StreamID into a base36 url
   */
  @Memoize()
  toUrl(): string {
    return `ceramic://${this.toString()}`
  }

  /**
   * If genesis commit:
   * CommitID(k3y52l7mkcvtg023bt9txegccxe1bah8os3naw5asin3baf3l3t54atn0cuy98yws, 0)
   *
   * If commit:
   * CommitID(k3y52l7mkcvtg023bt9txegccxe1bah8os3naw5asin3baf3l3t54atn0cuy98yws, bagjqcgzaday6dzalvmy5ady2m5a5legq5zrbsnlxfc2bfxej532ds7htpova)
   *
   * @returns {String}
   */
  [Symbol.for('nodejs.util.inspect.custom')](): string {
    return `CommitID(${this.toString()})`
  }

  /**
   * String representation of CommitID.
   */
  [Symbol.toPrimitive](): string | Uint8Array {
    return this.toString()
  }
}
