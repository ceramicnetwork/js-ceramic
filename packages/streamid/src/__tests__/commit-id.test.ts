import { jest } from '@jest/globals'
import * as util from 'util'
import { CommitID } from '../commit-id.js'
import { CID } from 'multiformats/cid'
import { base36 } from 'multiformats/bases/base36'

const BASE_CID_STRING = 'bagcqcerakszw2vsovxznyp5gfnpdj4cqm2xiv76yd24wkjewhhykovorwo6a'
const BASE_CID = CID.parse(BASE_CID_STRING)
const COMMIT_CID_STRING = 'bagjqcgzaday6dzalvmy5ady2m5a5legq5zrbsnlxfc2bfxej532ds7htpova'
const COMMIT_CID = CID.parse(COMMIT_CID_STRING)

const STREAM_ID_STRING = 'kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s'
const STREAM_ID_BYTES = base36.decode(STREAM_ID_STRING)
const STREAM_ID_URL = 'ceramic://kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s'
const STREAM_ID_URL_LEGACY =
  '/ceramic/kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s'

const STREAM_ID_WITH_COMMIT =
  'k1dpgaqe3i64kjqcp801r3sn7ysi5i0k7nxvs7j351s7kewfzr3l7mdxnj7szwo4kr9mn2qki5nnj0cv836ythy1t1gya9s25cn1nexst3jxi5o3h6qprfyju'
const STREAM_ID_WITH_COMMIT_BYTES = base36.decode(STREAM_ID_WITH_COMMIT)

const STREAM_ID_WITH_0_COMMIT = 'k3y52l7qbv1frxwipl4hp7e6jlu4f6u8upm2xv0irmedfkm5cnutmezzi3u7mytj4'
const STREAM_ID_WITH_0_COMMIT_BYTES = base36.decode(STREAM_ID_WITH_0_COMMIT)

const STREAM_ID_WITH_COMMIT_LEGACY =
  '/ceramic/kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s?commit=bagjqcgzaday6dzalvmy5ady2m5a5legq5zrbsnlxfc2bfxej532ds7htpova'
const STREAM_ID_WITH_0_COMMIT_LEGACY =
  '/ceramic/kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s?commit=0'

describe('constructor', () => {
  test('create by parts (type:int, cid:cid)', () => {
    const type = 0
    const streamid = new CommitID(type, BASE_CID)

    expect(streamid.type).toEqual(type)
    expect(streamid.cid).toEqual(BASE_CID)
    expect(streamid.toString()).toMatchSnapshot()
  })

  test('create by parts (type:string, cid:string, commit:null)', () => {
    const type = 'tile'
    const streamid = new CommitID(type, BASE_CID_STRING, null)

    expect(streamid.type).toEqual(0)
    expect(streamid.cid).toEqual(BASE_CID)
    expect(streamid.commit).toEqual(BASE_CID)
    expect(streamid.toString()).toMatchSnapshot()
  })

  test('create by parts (type:string, cid:string, commit:string)', () => {
    const type = 'tile'
    const streamid = new CommitID(type, BASE_CID_STRING, BASE_CID_STRING)

    expect(streamid.type).toEqual(0)
    expect(streamid.cid).toEqual(BASE_CID)
    expect(streamid.commit).toEqual(BASE_CID)
    expect(streamid.toString()).toMatchSnapshot()
  })

  test('unknown streamtype name', () => {
    expect(() => new CommitID('invalid-garbage', BASE_CID_STRING, BASE_CID_STRING)).toThrow()
  })
})

describe('#atCommit', () => {
  const commitId = new CommitID('tile', BASE_CID, COMMIT_CID)
  const streamId = commitId.baseID

  test('to number 0', () => {
    const traveller = CommitID.make(streamId, 0)
    expect(traveller.commit).toEqual(BASE_CID)
  })
  test('to number 1', () => {
    expect(() => CommitID.make(streamId, 1)).toThrowErrorMatchingSnapshot()
  })
  test('to commit CID', () => {
    const commitId = new CommitID('tile', BASE_CID)
    const traveller = CommitID.make(commitId.baseID, COMMIT_CID)
    expect(traveller.commit).toEqual(COMMIT_CID)
  })
  test('to commit CID as string', () => {
    const commitId = new CommitID('tile', BASE_CID)
    const traveller = CommitID.make(commitId.baseID, COMMIT_CID_STRING)
    expect(traveller.commit).toEqual(COMMIT_CID)
  })
  test('to garbage string', () => {
    expect(() => CommitID.make(streamId, 'garbage')).toThrow()
  })
})

describe('.fromBytes', () => {
  test('create from stream id bytes: failure', () => {
    expect(() => CommitID.fromBytes(STREAM_ID_BYTES)).toThrow()
  })

  test('invalid stream id', () => {
    expect(() => CommitID.fromBytes(BASE_CID.bytes)).toThrow()
  })

  test('create from bytes inlcuding commit', () => {
    const streamid = CommitID.fromBytes(STREAM_ID_WITH_COMMIT_BYTES)

    expect(streamid.type).toEqual(0)
    expect(streamid.cid).toEqual(BASE_CID)
    expect(streamid.commit).toEqual(COMMIT_CID)
    expect(streamid.toString()).toMatchSnapshot()
  })

  test('create from bytes inlcuding commit 0', () => {
    const streamid = CommitID.fromBytes(STREAM_ID_WITH_0_COMMIT_BYTES)

    expect(streamid.type).toEqual(0)
    expect(streamid.cid).toEqual(BASE_CID)
    expect(streamid.commit).toEqual(BASE_CID)
    expect(streamid.toString()).toMatchSnapshot()
  })

  test('roundtrip streamID bytes', () => {
    const commitId = new CommitID('tile', BASE_CID_STRING)
    const commitId2 = CommitID.fromBytes(commitId.bytes)
    expect(commitId.toString()).toEqual(commitId2.toString())
    expect(commitId.bytes).toEqual(commitId2.bytes)
  })

  test('roundtrip streamID bytes with commit', () => {
    const streamid = new CommitID('tile', BASE_CID_STRING, COMMIT_CID_STRING)
    const streamid2 = CommitID.fromBytes(streamid.bytes)
    expect(streamid.toString()).toEqual(streamid2.toString())
    expect(streamid.bytes).toEqual(streamid2.bytes)
  })

  test('roundtrip streamID bytes with commit', () => {
    const streamid = new CommitID('tile', BASE_CID_STRING, 0)
    const streamid2 = CommitID.fromBytes(streamid.bytes)
    expect(streamid.toString()).toEqual(streamid2.toString())
    expect(streamid.bytes).toEqual(streamid2.bytes)
  })
})

describe('.fromString', () => {
  test('create from stream id string: fail', () => {
    expect(() => CommitID.fromString(STREAM_ID_STRING)).toThrow()
  })

  test('create from string including commit', () => {
    const streamid = CommitID.fromString(STREAM_ID_WITH_COMMIT)

    expect(streamid.type).toEqual(0)
    expect(streamid.cid).toEqual(BASE_CID)
    expect(streamid.commit).toEqual(COMMIT_CID)
    expect(streamid.toString()).toMatchSnapshot()
  })

  test('create from string including commit 0', () => {
    const streamid = CommitID.fromString(STREAM_ID_WITH_0_COMMIT)

    expect(streamid.type).toEqual(0)
    expect(streamid.cid).toEqual(BASE_CID)
    expect(streamid.commit).toEqual(BASE_CID)
    expect(streamid.toString()).toMatchSnapshot()
  })

  test('create from stream id url string: fail', () => {
    expect(() => CommitID.fromString(STREAM_ID_URL)).toThrow()
  })

  test('create from legacy streamid string "/ceramic/": fail', () => {
    expect(() => CommitID.fromString(STREAM_ID_URL_LEGACY))
  })

  test('create from legacy string "/ceramic/" with commit param "?commit="', () => {
    const streamid = CommitID.fromString(STREAM_ID_WITH_COMMIT_LEGACY)

    expect(streamid.type).toEqual(0)
    expect(streamid.cid).toEqual(BASE_CID)
    expect(streamid.commit).toEqual(COMMIT_CID)
    expect(streamid.toString()).toMatchSnapshot()
  })

  test('create from legacy string "/ceramic/" with commit param "?commit=0"', () => {
    const streamid = CommitID.fromString(STREAM_ID_WITH_0_COMMIT_LEGACY)

    expect(streamid.type).toEqual(0)
    expect(streamid.cid).toEqual(BASE_CID)
    expect(streamid.commit).toEqual(BASE_CID)
    expect(streamid.toString()).toMatchSnapshot()
  })

  test('roundtrip streamID string', () => {
    const streamid = new CommitID('tile', BASE_CID_STRING)
    const streamid2 = CommitID.fromString(streamid.toString())
    expect(streamid.toString()).toEqual(streamid2.toString())
  })

  test('roundtrip streamID string with commit', () => {
    const streamid = new CommitID('tile', BASE_CID_STRING, COMMIT_CID_STRING)
    const streamid2 = CommitID.fromString(streamid.toString())
    expect(streamid.toString()).toEqual(streamid2.toString())
  })

  test('roundtrip streamID string with commit 0', () => {
    const streamid = new CommitID('tile', BASE_CID_STRING, 0)
    const streamid2 = CommitID.fromString(streamid.toString())
    expect(streamid.toString()).toEqual(streamid2.toString())
  })
})

test('.bytes', () => {
  const streamid = new CommitID('tile', BASE_CID_STRING)
  const bytes = streamid.bytes
  expect(bytes).toBeDefined()
  expect(bytes instanceof Uint8Array).toEqual(true)
  expect(bytes).toMatchSnapshot()
})

test('.typeName if registered', () => {
  const streamid = new CommitID('tile', BASE_CID_STRING, null)
  expect(streamid.typeName).toEqual('tile')
  const streamid2 = new CommitID(10, BASE_CID_STRING)
  expect(() => streamid2.typeName).toThrowErrorMatchingSnapshot()
})

test('.toString()', () => {
  const streamid = new CommitID('tile', BASE_CID_STRING)
  const str = streamid.toString()
  expect(str).toBeDefined()
  expect(str).toMatchSnapshot()
})

test('.toUrl()', () => {
  const streamid = new CommitID('tile', BASE_CID_STRING)
  const str = streamid.toUrl()
  expect(str).toBeDefined()
  expect(str.includes('ceramic://')).toEqual(true)
  expect(str).toMatchSnapshot()
})

describe('#equals', () => {
  test('equals other streamID', () => {
    const streamid = new CommitID('tile', BASE_CID_STRING)
    const streamid2 = new CommitID('tile', BASE_CID_STRING)
    const streamid3 = new CommitID('caip10-link', BASE_CID_STRING)
    expect(streamid.equals(streamid2)).toEqual(true)
    expect(streamid.equals(streamid3)).toEqual(false)
  })
})

describe('nodejs inspect', () => {
  test('genesis commit', () => {
    const streamid = new CommitID('tile', BASE_CID_STRING)
    expect(util.inspect(streamid)).toMatchSnapshot()
  })
  test('commit', () => {
    const streamid = new CommitID('tile', BASE_CID_STRING, COMMIT_CID_STRING)
    expect(util.inspect(streamid)).toMatchSnapshot()
  })
})

test('to primitive', () => {
  const streamid = new CommitID('tile', BASE_CID_STRING, COMMIT_CID_STRING)
  expect(`${streamid}`).toEqual(streamid.toString())
  expect(+streamid).toBeNaN()
  expect(streamid + '').toEqual(streamid.toString())
})

test('#baseID', () => {
  const commitId = new CommitID('tile', BASE_CID_STRING, COMMIT_CID_STRING)
  const streamId = commitId.baseID
  expect(streamId.type).toEqual(commitId.type)
  expect(streamId.cid).toEqual(commitId.cid)
})

test('instanceof', () => {
  const commitId = new CommitID('tile', BASE_CID_STRING, COMMIT_CID_STRING)
  const instanceSpy = jest.spyOn(CommitID, 'isInstance')
  expect(CommitID.isInstance(commitId)).toBeTruthy()
  expect(instanceSpy).toBeCalledWith(commitId)
})
