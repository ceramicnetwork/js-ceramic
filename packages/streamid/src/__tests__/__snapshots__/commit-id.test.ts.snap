// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#atCommit to number 1 1`] = `"Cannot specify commit as a number except to request commit 0 (the genesis commit)"`;

exports[`.bytes 1`] = `
Uint8Array [
  206,
  1,
  0,
  1,
  133,
  1,
  18,
  32,
  84,
  179,
  109,
  86,
  78,
  173,
  242,
  220,
  63,
  166,
  43,
  94,
  52,
  240,
  80,
  102,
  174,
  138,
  255,
  216,
  30,
  185,
  101,
  36,
  150,
  57,
  240,
  167,
  85,
  209,
  179,
  188,
  0,
]
`;

exports[`.fromBytes create from bytes inlcuding commit 0 1`] = `"k3y52l7qbv1frxwipl4hp7e6jlu4f6u8upm2xv0irmedfkm5cnutmezzi3u7mytj4"`;

exports[`.fromBytes create from bytes inlcuding commit 1`] = `"k1dpgaqe3i64kjqcp801r3sn7ysi5i0k7nxvs7j351s7kewfzr3l7mdxnj7szwo4kr9mn2qki5nnj0cv836ythy1t1gya9s25cn1nexst3jxi5o3h6qprfyju"`;

exports[`.fromString create from legacy string "/ceramic/" with commit param "?commit=" 1`] = `"k1dpgaqe3i64kjqcp801r3sn7ysi5i0k7nxvs7j351s7kewfzr3l7mdxnj7szwo4kr9mn2qki5nnj0cv836ythy1t1gya9s25cn1nexst3jxi5o3h6qprfyju"`;

exports[`.fromString create from legacy string "/ceramic/" with commit param "?commit=0" 1`] = `"k3y52l7qbv1frxwipl4hp7e6jlu4f6u8upm2xv0irmedfkm5cnutmezzi3u7mytj4"`;

exports[`.fromString create from string including commit 0 1`] = `"k3y52l7qbv1frxwipl4hp7e6jlu4f6u8upm2xv0irmedfkm5cnutmezzi3u7mytj4"`;

exports[`.fromString create from string including commit 1`] = `"k1dpgaqe3i64kjqcp801r3sn7ysi5i0k7nxvs7j351s7kewfzr3l7mdxnj7szwo4kr9mn2qki5nnj0cv836ythy1t1gya9s25cn1nexst3jxi5o3h6qprfyju"`;

exports[`.toString() 1`] = `"k3y52l7qbv1frxwipl4hp7e6jlu4f6u8upm2xv0irmedfkm5cnutmezzi3u7mytj4"`;

exports[`.toUrl() 1`] = `"ceramic://k3y52l7qbv1frxwipl4hp7e6jlu4f6u8upm2xv0irmedfkm5cnutmezzi3u7mytj4"`;

exports[`.typeName if registered 1`] = `"No stream type registered for index 10"`;

exports[`constructor create by parts (type:int, cid:cid) 1`] = `"k3y52l7qbv1frxwipl4hp7e6jlu4f6u8upm2xv0irmedfkm5cnutmezzi3u7mytj4"`;

exports[`constructor create by parts (type:string, cid:string, commit:null) 1`] = `"k3y52l7qbv1frxwipl4hp7e6jlu4f6u8upm2xv0irmedfkm5cnutmezzi3u7mytj4"`;

exports[`constructor create by parts (type:string, cid:string, commit:string) 1`] = `"k3y52l7qbv1frxwipl4hp7e6jlu4f6u8upm2xv0irmedfkm5cnutmezzi3u7mytj4"`;

exports[`nodejs inspect commit 1`] = `"CommitID(k1dpgaqe3i64kjqcp801r3sn7ysi5i0k7nxvs7j351s7kewfzr3l7mdxnj7szwo4kr9mn2qki5nnj0cv836ythy1t1gya9s25cn1nexst3jxi5o3h6qprfyju)"`;

exports[`nodejs inspect genesis commit 1`] = `"CommitID(k3y52l7qbv1frxwipl4hp7e6jlu4f6u8upm2xv0irmedfkm5cnutmezzi3u7mytj4)"`;
